src/typescript.js

```diff
@@ -0,0 +1,110 @@
+// @ts-check
+
+// @ts-expect-error -- False positive
+import importPlugin from 'eslint-plugin-import';
+// eslint-disable-next-line import/no-unresolved -- False positive
+import typescriptEslint from 'typescript-eslint';
+import { ERROR, OFF, typescriptFiles } from './config.js';
+import javascript from './javascript.js';
+
+/**
+ * @type {Array<import('eslint').Linter.Config>}
+ */
+export default [
+  ...javascript,
+
+  typescriptEslint.configs.eslintRecommended,
+  ...typescriptEslint.configs.recommended,
+  importPlugin.configs.typescript,
+
+  {
+    // Only apply TypeScript rules to TypeScript files to avoid
+    // causing issues in regular JavaScript files. See also:
+    // https://stackoverflow.com/a/64488474/343045
+    files: typescriptFiles,
+    settings: {
+      'import/resolver': 'typescript'
+    },
+    languageOptions: {
+      parserOptions: {
+        projectService: true,
+        tsconfigRootDir: import.meta.dirname
+      }
+    },
+    rules: {
+      // The TypeScript compiler takes care of this
+      'import/no-unresolved': OFF,
+      'import/named': OFF,
+      'import/namespace': OFF,
+      'import/default': OFF,
+      'import/export': OFF,
+
+      // Use the TypeScript alternatives
+      'no-unused-vars': OFF,
+      'no-shadow': OFF,
+      'no-use-before-define': OFF,
+      'no-unused-expressions': OFF,
+
+      '@typescript-eslint/no-unused-vars': [
+        ERROR,
+        {
+          args: 'after-used',
+          argsIgnorePattern: '^_',
+          vars: 'all'
+        }
+      ],
+      '@typescript-eslint/no-unused-expressions': ERROR,
+      '@typescript-eslint/array-type': [ERROR, { default: 'generic' }],
+      '@typescript-eslint/await-thenable': ERROR,
+      '@typescript-eslint/ban-ts-comment': [
+        ERROR,
+        {
+          'ts-expect-error': false, // This is sometimes necessary and a better alternative to ts-ignore
+          'ts-ignore': 'allow-with-description'
+        }
+      ],
+      // Avoid declaring the implied return type for React components
+      '@typescript-eslint/explicit-module-boundary-types': OFF,
+      '@typescript-eslint/method-signature-style': [ERROR, 'property'],
+      '@typescript-eslint/no-explicit-any': ERROR,
```

Comment:
`no-explicit-any` is already enabled in `recommended` and can be removed here (see https://github.com/tools-aoeur/component-library/pull/1954#discussion_r1796142930)).